nntrainer/include/layer_internal.h

```diff
@@ -213,7 +213,7 @@ class Layer : public ml::train::Layer {
    * @brief     set trainable for this layer
    * @param[in] train to enable/disable train
    */
-  void setTrainable(bool train) noexcept { trainable = train; }
+  virtual void setTrainable(bool train) { trainable = train; }
```

Comment:
Could you elaborate the trainable?? Like conv, fc are trainable layers and input layer is not. Or does it means we are able to set conv with not-trainable?? For the nnstreamer layer, it is not trainable I guess. but if we are using nntrainer ini as a backbone, we can train. In this case, is it depends on user to set trainable or untrainable?? I think we haven't decided how to use this variable.

Reply:
> For the nnstreamer layer, it is not trainable I guess. but if we are using nntrainer ini as a backbone, we can train. In this case, is it depends on user to set trainable or untrainable?? 

trainable is by default set to true for a layer created with nntrainer ini file.
the exception is backbone ini, where the default trainable is false (must be set true with `trainable` property of the backbone in ini or API).
For external framework layers (tflite/nnstreamer), setting trainable to true will throw exception, setting to false is only allowed.

> Could you elaborate the trainable?? Like conv, fc are trainable layers and input layer is not. Or does it means we are able to set conv with not-trainable?? 

Yes, this property can be set individually for the layer from ini. There is no limitation to the layers which can be set to trainable/non-trainable. Setting a layer to trainable does not essentially mean that that layer has some variables which are going to be trained (flatten layer allows setting trainable to true/false but does not effect the behavior of the model training/inference).

Reply:
> Setting a layer to trainable does not essentially mean that that layer has some variables which are going to be trained (flatten layer allows setting trainable to true/false but does not effect the behavior of the model training/inference).
So, if user set trainable as false for conv layer, it just skip update weight gradient. I think we haven't implement this so far. But we are going to. Am I right?

Reply:
>So, if user set trainable as false for conv layer, it just skip update weight gradient. I think we haven't implement this so far. But we are going to. Am I right?

We have implemented this. We are skipping updating the gradient (in Optimizer::apply_gradients)

Reply:
IMO also this is already supported. In some cases where its not supported, I will add it :)

Reply:
@jijoongmoon Currently we only conv2d, fc and bn layer which updates weights.
Check line 353 for conv2d_layer.cpp, 129 for fc_layer.cpp. It is missing for bn_layer, which will be added soon.

Reply:
I'll add if you are okay

Reply:
@zhoonit Thanks but I already added.